### ISSUE TITLE
[Bug]: Prompt "Return a JSON object that uses the SAY command to say what you're thinking." sent to LLM for no apparent reason

### ISSUE THREAD

**Issue by dclaux** [microsoft/teams-ai#1130]
### Language

Javascript/Typescript

### Version

1.0.1

### Description

The library is sending an undesired/useless "Return a JSON object that uses the SAY command to say what you're thinking." prompt to the LLM that leads to poor results.

I have a Bot whose prompt file is defined as follows:

```
The following is a conversation with an AI assistant.
The assistant is an expert at designing Adaptive Cards.
Any changes made to the card must comply with the Adaptive Card Schema.

context:
The current card payload is {{getCardPayload}}.
```

**Update**: the function call in the prompt file seems to be what causes the issue. If I hard code something in there instead of calling the function, things work as expected.

The `getCardPayload` function is added to my planner as follows:

![image](https://github.com/microsoft/teams-ai/assets/1334689/e9fa55cb-e1c2-474a-909e-79d00291b332)

Sending the prompt "What is the current card payload?" ends up prompting the LLM to create a plan that includes a single SAY command with the response I'd expect ("The current card payload is ...").

But, somehow `teams-ai` decides to prompt the LLM again with
```
{
    "role": "user",
    "content": "Return a JSON object that uses the SAY command to say what you're thinking."
}
```

which leads the LLM to return "I am an AI assistant and I am thinking about Adaptive Cards."

See log in repro steps below.

### Reproduction Steps

```bash
CHAT PROMPT:
[
  {
    "role": "user",
    "content": "The following is a conversation with an AI assistant.\nThe assistant is an expert at designing Adaptive Cards.\nAny changes made to the card must comply with the Adaptive Card Schema.\n\ncontext:\nThe current card payload is {\"type\":\"AdaptiveCard\",\"version\":\"1.0\",\"body\":[{\"type\":\"TextBlock\",\"text\":\"Hello World!\"}]}.\n\nactions:\n  setCardPayload:\n    description: Sets the current card payload\n    parameters:\n      payload:\n        type: object\n        description: The updated card payload\n\n\nUse the actions above to create a plan in the following JSON format:\n{\"type\":\"plan\",\"commands\":[{\"type\":\"DO\",\"action\":\"<name>\",\"parameters\":{\"<name>\":<value>}},{\"type\":\"SAY\",\"response\":\"<response>\"}]}"
  },
  {
    "role": "user",
    "content": "What is the current card payload?"
  }
]
CHAT RESPONSE:
status: 200
duration: 1252ms
{
  "id": "chatcmpl-8dU5SbAipcWUYw9VqDTRXeayElhi4",
  "object": "chat.completion",
  "created": 1704421138,
  "model": "gpt-35-turbo-16k",
  "prompt_filter_results": [
    {
      "prompt_index": 0,
      "content_filter_results": {
        "hate": {
          "filtered": false,
          "severity": "safe"
        },
        "self_harm": {
          "filtered": false,
          "severity": "safe"
        },
        "sexual": {
          "filtered": false,
          "severity": "safe"
        },
        "violence": {
          "filtered": false,
          "severity": "safe"
        }
      }
    }
  ],
  "choices": [
    {
      "finish_reason": "stop",
      "index": 0,
      "message": {
        "role": "assistant",
        "content": "{\"type\":\"plan\",\"commands\":[{\"type\":\"SAY\",\"response\":\"The current card payload is {\"type\":\"AdaptiveCard\",\"version\":\"1.0\",\"body\":[{\"type\":\"TextBlock\",\"text\":\"Hello World!\"}]}.\"}]}"
      },
      "content_filter_results": {
        "hate": {
          "filtered": false,
          "severity": "safe"
        },
        "self_harm": {
          "filtered": false,
          "severity": "safe"
        },
        "sexual": {
          "filtered": false,
          "severity": "safe"
        },
        "violence": {
          "filtered": false,
          "severity": "safe"
        }
      }
    }
  ],
  "usage": {
    "prompt_tokens": 175,
    "completion_tokens": 49,
    "total_tokens": 224
  }
}
CHAT PROMPT:
[
  {
    "role": "user",
    "content": "The following is a conversation with an AI assistant.\nThe assistant is an expert at designing Adaptive Cards.\nAny changes made to the card must comply with the Adaptive Card Schema.\n\ncontext:\nThe current card payload is {\"type\":\"AdaptiveCard\",\"version\":\"1.0\",\"body\":[{\"type\":\"TextBlock\",\"text\":\"Hello World!\"}]}.\n\nactions:\n  setCardPayload:\n    description: Sets the current card payload\n    parameters:\n      payload:\n        type: object\n        description: The updated card payload\n\n\nUse the actions above to create a plan in the following JSON format:\n{\"type\":\"plan\",\"commands\":[{\"type\":\"DO\",\"action\":\"<name>\",\"parameters\":{\"<name>\":<value>}},{\"type\":\"SAY\",\"response\":\"<response>\"}]}"
  },
  {
    "role": "user",
    "content": "What is the current card payload?"
  },
  {
    "role": "assistant",
    "content": "{\"type\":\"plan\",\"commands\":[{\"type\":\"SAY\",\"response\":\"The current card payload is {\"type\":\"AdaptiveCard\",\"version\":\"1.0\",\"body\":[{\"type\":\"TextBlock\",\"text\":\"Hello World!\"}]}.\"}]}"
  },
  {
    "role": "user",
    "content": "Return a JSON object that uses the SAY command to say what you're thinking."
  }
]
CHAT RESPONSE:
status: 200
duration: 643ms
{
  "id": "chatcmpl-8dU5TS3T3lv11UKwyoczVs6YZyRCq",
  "object": "chat.completion",
  "created": 1704421139,
  "model": "gpt-35-turbo-16k",
  "prompt_filter_results": [
    {
      "prompt_index": 0,
      "content_filter_results": {
        "hate": {
          "filtered": false,
          "severity": "safe"
        },
        "self_harm": {
          "filtered": false,
          "severity": "safe"
        },
        "sexual": {
          "filtered": false,
          "severity": "safe"
        },
        "violence": {
          "filtered": false,
          "severity": "safe"
        }
      }
    }
  ],
  "choices": [
    {
      "finish_reason": "stop",
      "index": 0,
      "message": {
        "role": "assistant",
        "content": "{\"type\":\"plan\",\"commands\":[{\"type\":\"SAY\",\"response\":\"I am an AI assistant and I am thinking about Adaptive Cards.\"}]}"
      },
      "content_filter_results": {
        "hate": {
          "filtered": false,
          "severity": "safe"
        },
        "self_harm": {
          "filtered": false,
          "severity": "safe"
        },
        "sexual": {
          "filtered": false,
          "severity": "safe"
        },
        "violence": {
          "filtered": false,
          "severity": "safe"
        }
      }
    }
  ],
  "usage": {
    "prompt_tokens": 248,
    "completion_tokens": 29,
    "total_tokens": 277
  }
}
```

**Comment by BlueWhaleSEO** [comment]
I've removed all the functions from the planner and I'm still getting the same extra line `"Return a JSON object that uses the SAY command to say what you're thinking."`. 

Could that be why I'm getting JSON plan, when calling `planner.completePrompt()` instead of [now supposedly removed](https://github.com/microsoft/teams-ai/discussions/1060#discussioncomment-8117891) `app.ai.chain()`?

**Comment by donatas-xyz** [comment]
Hi @dclaux @aacebo 

Anything new in regards of this bug? 

I still can't progress in any meaningful way with the version 1.*.* because of this bug. I find it hard to believe, that no one else is having it? 

Thank you!

**Comment by simaonogueira101** [comment]
Hi @donatas-xyz, @dclaux, @aacebo and @singhk97,

I came across this issue as well. It seems the JSONResponseValidator (js/packages/teams-ai/src/validators/JSONResponseValidator.ts) is incorrectly invalidating the text if it includes line breaks. The parseJSON function in parseAllObjects that is imported from Response (js/packages/teams-ai/src/validators/Response.ts) does not account for these line breaks and therefore returns undefined when trying JSON.parse.

A suggestion to fix the issue (bearing in mind that I did not test this thoroughly) would be to add `text = text.replace(/\n/g, "\\n").replace(/\r/g, "\\r");` as the first line in `static parseJSON(text) {`. It does solve this issue in specific.

In the meantime, without the proposed changes to the package, I've added the following lines at the end of my skprompt.txt:
```
Your answers MUST be in an easy to read format with paragraphs, line breaks, dashed lists and bold text.
You MUST replace all "\n", "\r" and "\t" line breaks in your answer with "((LB))".
```

Afterwards I use (was already using before this fix to create AdaptiveCards) an action targetting the SAY command to format the response and convert back to line breaks. **This solution pollutes the chat history with "((LB))" so a better fix is in order.**

This problem appears to be the same as the one in https://github.com/microsoft/teams-ai/issues/1245. @singhk97 your suggestion to use `json_mode` did not fix my issue as the problem is not with generating JSON but with parsing it.

Nevertheless, my solution is working quite well on my use case and so I'll leave it for now. I do hope the team manages to fix this issue and I hope this comment helps you as well!

Best regards,
Simão

**Comment by donatas-xyz** [comment]
Hi @simaonogueira101 

Thank you for your response. 

Perhaps I'm not doing it the right way, but I've edited `C:\Users\myuser\source\ChatBotApp\node_modules\@microsoft\teams-ai\lib\validators\Response.js` and added your proposed code change to it, deployed it back to Azure, however I'm still having the same issue? Or should I edit `Response.ts` file deployed in Azure? 

Thank you once again @simaonogueira101 - it's good to know that at least someone has found a solution to this issue :)

**Comment by donatas-xyz** [comment]
Upgraded to teams-ai v1.3.1, but the issue is still there.

**Comment by donatas-xyz** [comment]
Ok, I've been adding functionality from the old chatbot app with this bug to a new chatbot app and checking if/when the issue appears again. 

I was able to replicate the issue. The short of it is that changing `config.json` schema from 1 to 1.1 immediately causes this bug. 

I haven't tried to manipulate `config.json` 1.1 version to see if there are any specific parameters causing it yet. 

Here are the reproduction steps: 

1. Created a basic bot app with teams-ai 1.0.0
2. Updated dependencies to their latest versions: 
```json
    "dependencies": {
        "@microsoft/teams-ai": "~1.3.1",
        "botbuilder": "^4.22.2",
        "restify": "~11.1.0",
        "shx": "^0.3.4"
    },
    "devDependencies": {
        "@types/restify": "8.5.12",
        "@types/node": "^20.14.0",
        "@types/debug": "^4.1.12",
        "debug": "^4.3.5",
        "env-cmd": "^10.1.0",
        "ts-node": "^10.9.2",
        "typescript": "^5.4.5",
        "nodemon": "~3.1.0"
    }
``` 
3. Updated the rest of package.json values , such as: 
```json
    "version": "1.1.2",
    "description": "Microsoft Teams Toolkit AI Chat Bot Sample",
    "engines": {
        "node": ">=18.0.0"
    },
    "author": "Donatas",
    "license": "UNLICENSED",
    "keywords": [
        "Bob",
        "chatbot",
        "bot"
    ]
```
4. `Debug` dependencies and code bits added (could not be used due to #1763 bug)
5. Added a basic function to return my name as a string to `skprompt.txt`. 
```typescript
export function helperFunctions(): void
{
    planner.prompts.addFunction('getUsersName', async (context: TurnContext, memory: Memory) => 
    {
        let usersFirstName: string = "Donatas";

        return usersFirstName;
    });
}
```
```text
#### START PROMPT INSTRUCTIONS

## Prompt grounding
- You are an AI assistant that helps people find information. 
- Your name is Bob. 
- The Person is someone who wants to perform actions or find out information. 
- The Person's name is {{getUsersName}}. 

#### END PROMPT INSTRUCTIONS

#### START ACTION INSTRUCTIONS 

What to do after action and functions have been executed: 
- Return the action results to the Person. 
- If the chat doesn't match the action then just use standard chat. 
- The Assistant must return the following JSON structure if it matches an action:
{"type":"plan","commands":[{"type":"DO","action":"<name>","entities":{"<name>":<value>}},{"type":"SAY","response":"<response>"}]}

#### END ACTION INSTRUCTIONS 

#### START ACTION EXAMPLES

Person: What is my name? 
Bob: Your name is {{getUsersName}} 

#### END ACTION EXAMPLES

{{$history}}
Person: {{$input}}
Bob: 
```
6. `Config.json` schema 1.0 - **everything still works**.
```json
{
  "schema": 1,
  "description": "AI Bot",
  "type": "completion",
  "completion": {
    "max_tokens": 1000,
    "temperature": 0.8,
    "top_p": 0.0,
    "presence_penalty": 0.6,
    "frequency_penalty": 0.0
  }
}
```
7. Changed `config.json` schema to 1.1. No `actions.json` file present at this point. **The bug appears immediately**. 
```json
{
	"schema": 1.1,
	"description": "AI Bot",
	"type": "completion",
	"completion": {
		"model": "gpt3_5-custom-azure-model",
		"completion_type": "chat",
		"include_history": true,
		"include_input": true,
		"max_input_tokens": 2800,
		"max_tokens": 1000,
		"temperature": 0.8,
		"top_p": 0.0,
		"presence_penalty": 0.6,
		"frequency_penalty": 0.0,
		"stop_sequences": [
			"Person:",
			"Bob:"
		]
	},
	"augmentation": {
		"augmentation_type": "sequence"
	}
}
``` 

Hopefully this will help to identify the root cause for this bug. 

FYI @dclaux @aacebo @singhk97 @simaonogueira101

**Comment by donatas-xyz** [comment]
I've done some more debugging and these are my findings: 

1. The reason why this bug appears when _sequence_ augmentation is enabled, is because of the logic in `@microsoft\teams-ai\lib\augmentations\SequenceAugmentation.js` file and class. Like I said, I'm not an expert, but my understanding is that the below command appearing in the bug is activated, when validator fails to validate the response from the assistant: 
```javascript
class SequenceAugmentation {
    _section;
    _planValidator = new validators_1.JSONResponseValidator(exports.PlanSchema, `Return a JSON object that uses the SAY command to say what you're thinking.`);
    _actionValidator;
    constructor(actions) {
        this._section = new ActionAugmentationSection_1.ActionAugmentationSection(actions, [
            `Use the actions above to create a plan in the following JSON format:`,
            `{"type":"plan","commands":[{"type":"DO","action":"<name>","parameters":{"<name>":<value>}},{"type":"SAY","response":"<response>"}]}`
        ].join('\n'));
        this._actionValidator = new validators_1.ActionResponseValidator(actions, true);
    }
}    
``` 
2. If we execute this simple scenario, the following will happen: 
a) Start the bot and ask a simple question: `"Who are you?"` 
b) The bot will correctly respond `"I am Bob, an AI assistant designed to help you find information and perform actions. How can I assist you today?"`. 
c) The assistant's response in step 2.b is then being passed to `@microsoft\teams-ai\lib\validators\JSONResponseValidator.js` with the custom feedback message (the bug) string from `SequenceAugmentation` class. This message is being used, when no JSON is returned. `missingJsonFeedback` is the bug's message or if it's not there, it would use `No valid JSON objects were found in the response. Return a valid JSON object.` instead. As an experiment, I've commented `missingJsonFeedback` parameter out, which forced the alternative message, which's `Return a valid JSON object` request executed a random action in the prompt...: 
```javascript
class JSONResponseValidator {
    /**
     * Creates a new `JSONResponseValidator` instance.
     * @param {Schema} schema Optional. JSON schema to validate the response against.
     * @param {string} missingJsonFeedback Optional. Custom feedback to give when no JSON is returned.
     * @param {string} errorFeedback Optional. Custom feedback prefix to use when schema errors are detected.
     */
    constructor(schema, missingJsonFeedback, errorFeedback) {
        this.schema = schema;
        this.missingJsonFeedback =
             missingJsonFeedback ?? 'No valid JSON objects were found in the response. Return a valid JSON object.';
        this.errorFeedback = errorFeedback ?? 'The JSON returned had errors. Apply these fixes:';
    }
}    
``` 
3. So now when we have the logic for when no JSON is returned - what makes it think there was no JSON returned in the first place? It would appear that is caused by the logic in `JSONResponseValidator.validateResponse()` metthod. More specifically it's in `@microsoft\teams-ai\lib\validators\Response.js` `parseAllObjects` method. @simaonogueira101 mentioned it in his post above, but I believe the issue is that `parseAllObjects` seems to expect a pure JSON string to be passed in, when the actual is string it's getting is the one in step 2.b `I am Bob, an AI assistant designed to help you find information and perform actions. How can I assist you today?`: 
```javascript
class Response {
    /**
     * Parse all objects from a response string.
     * @param {string} text Response text to parse.
     * @returns {Record<string, any>[]} Array of parsed objects.
     */
    static parseAllObjects(text) {
        // First try parsing each line
        const objects = [];
        const lines = text.split('\n');
        if (lines.length > 1) {
            for (let i = 0; i < lines.length; i++) {
                const line = lines[i];
                const obj = this.parseJSON(line);
                if (obj) {
                    objects.push(obj);
                }
            }
        }
        // Next try parsing the entire text
        if (objects.length == 0) {
            const obj = this.parseJSON(text);
            if (obj) {
                objects.push(obj);
            }
        }
        return objects;
    }
}
``` 
Therefore whether it has line breaks or not - it doesn't even matter, because `this.parseJSON(text)` cannot process non-JSON strings and returns `undefined` instead. This makes `parseAllObjects` method return an empty array as well.  

4. When an array from `parseAllObjects` is getting returned inside `JSONResponseValidator.validateResponse()`, its checking if the resulting array is empty, and if it is - the validation fails and this then results in model trying to fix the response with the buggy instructions from `SequenceAugmentation` logic: 
```javascript
validateResponse(context, memory, tokenizer, response, remaining_attempts) {
        const message = response.message;
        const text = message.content ?? '';
        // Parse the response text
        const parsed = Response_1.Response.parseAllObjects(text);
        if (parsed.length == 0) {
            return Promise.resolve({
                type: 'Validation',
                valid: false,
                feedback: this.missingJsonFeedback
            });
        }
}
```
I hope this helps. 

FYI @dclaux @aacebo @singhk97

**Comment by donatas-xyz** [comment]
I've also created a new basic AI chatbot app using Teams Toolkit, but this time I've ran above configuration against public OpenAI model instead of Azure OpenAI model. **It fails** with the same error as soon as _sequence_ augmentation is enabled! 

Bot apps are running against ChatGPT 3.5 model. 

How is it possible, that only a couple of other people have had the same issue ?? 

@dclaux @simaonogueira101 @aacebo @singhk97 @corinagum

**Comment by aacebo** [comment]
hello @donatas-xyz, that logic is in place to enable the usage of our actions, but we are currently moving towards using tools for this instead. Our next minor release will include tools support for openai models and will fix this issue.

**Comment by donatas-xyz** [comment]
Thank you @aacebo. It would be fantastic as this issue is an absolute show stopper for me at least.

**Comment by packet3** [comment]
I am getting the exact same error message,  but I am using C# with Azure OpenAI.  any news when this will be resolved as I can't really continue.

```
The bot encountered an unhandled error: Reached max model response repair attempts. Last feedback given to model: "Return a JSON object that uses the SAY command to say what you're thinking."
```

this is my stack trace

```
Microsoft.Teams.AI.TeamsAdapter: Information: SendActivitiesAsync for 1 activities.
Microsoft.Teams.AI.TeamsAdapter: Information: Sending activity.  ReplyToId: 1720538447636
Microsoft.Teams.AI.TeamsAdapter: Information: SendActivitiesAsync for 1 activities.
Microsoft.Teams.AI.TeamsAdapter: Information: Sending activity.  ReplyToId: 1720538447636
Microsoft.Teams.AI.AI.Models.OpenAIModel: Trace: RESPONSE:
Microsoft.Teams.AI.AI.Models.OpenAIModel: Trace: status 200
Microsoft.Teams.AI.AI.Models.OpenAIModel: Trace: duration 1787.3755 ms
Microsoft.Teams.AI.AI.Models.OpenAIModel: Trace: {
  "Id": "chatcmpl-9j6w5Xs1fVD8V0ZyElVk8TBrA5WOr",
  "Created": "2024-07-09T15:20:49+00:00",
  "Choices": [
    {
      "Message": {
        "Role": {},
        "Content": "I can turn the lights on or off, check the current status of the lights, and pause for a specified amount of time. Here is a summary of the actions I support:\n\n{\"Actions\":{\"LightsOn\":{\"Description\":\"Turns on the lights\",\"Parameters\":null},\"LightsOff\":{\"Description\":\"Turns off the lights\",\"Parameters\":null},\"LightStatus\":{\"Description\":\"Gets the lights status\",\"Parameters\":null},\"Pause\":{\"Description\":\"Delays for a period of time\",\"Parameters\":{\"type\":\"object\",\"properties\":{\"time\":{\"type\":\"number\",\"description\":\"The amount of time to delay in milliseconds\"}},\"required\":[\"time\"]}}}}\n\nYou can use these actions to create a plan in the following JSON format:\n{\"type\":\"plan\",\"commands\":[{\"type\":\"DO\",\"action\":\"<name>\",\"parameters\":{\"<name>\":<value>}},{\"type\":\"SAY\",\"response\":\"<response>\"}]}",
        "ToolCalls": [],
        "FunctionCall": null,
        "AzureExtensionsContext": null
      },
      "LogProbabilityInfo": null,
      "Index": 0,
      "FinishReason": {},
      "FinishDetails": null,
      "ContentFilterResults": {
        "Sexual": {
          "Filtered": false,
          "Severity": {}
        },
        "Violence": {
          "Filtered": false,
          "Severity": {}
        },
        "Hate": {
          "Filtered": false,
          "Severity": {}
        },
        "SelfHarm": {
          "Filtered": false,
          "Severity": {}
        },
        "Profanity": null,
        "CustomBlocklists": null,
        "Error": null,
        "ProtectedMaterialText": null,
        "ProtectedMaterialCode": null
      },
      "Enhancements": null
    }
  ],
  "Model": "gpt-35-turbo",
  "PromptFilterResults": [
    {
      "PromptIndex": 0,
      "ContentFilterResults": {
        "Sexual": {
          "Filtered": false,
          "Severity": {}
        },
        "Violence": {
          "Filtered": false,
          "Severity": {}
        },
        "Hate": {
          "Filtered": false,
          "Severity": {}
        },
        "SelfHarm": {
          "Filtered": false,
          "Severity": {}
        },
        "Profanity": null,
        "CustomBlocklists": null,
        "Error": null,
        "Jailbreak": null,
        "IndirectAttack": null
      }
    }
  ],
  "SystemFingerprint": null,
  "Usage": {
    "CompletionTokens": 181,
    "PromptTokens": 588,
    "TotalTokens": 769
  }
}
Microsoft.Teams.AI.AI.Clients.LLMClient: Information: Response Repair Failed
Microsoft.Teams.AI.TeamsAdapter: Error: [OnTurnError] unhandled error : Reached max model response repair attempts. Last feedback given to model: "Return a JSON object that uses the SAY command to say what you're thinking."

System.Exception: Reached max model response repair attempts. Last feedback given to model: "Return a JSON object that uses the SAY command to say what you're thinking."
   at Microsoft.Teams.AI.AI.Planners.ActionPlanner`1.ContinueTaskAsync(ITurnContext context, TState state, AI`1 ai, CancellationToken cancellationToken)
   at Microsoft.Teams.AI.AI.Planners.ActionPlanner`1.BeginTaskAsync(ITurnContext context, TState state, AI`1 ai, CancellationToken cancellationToken)
   at Microsoft.Teams.AI.AI.AI`1.RunAsync(ITurnContext turnContext, TState turnState, Nullable`1 startTime, Int32 stepCount, CancellationToken cancellationToken)
   at Microsoft.Teams.AI.Application`1._OnTurnAsync(ITurnContext turnContext, CancellationToken cancellationToken)
   at Microsoft.Teams.AI.Application`1.OnTurnAsync(ITurnContext turnContext, CancellationToken cancellationToken)
   at Microsoft.Bot.Builder.MiddlewareSet.ReceiveActivityWithStatusAsync(ITurnContext turnContext, BotCallbackHandler callback, CancellationToken cancellationToken)
   at Microsoft.Bot.Builder.BotAdapter.RunPipelineAsync(ITurnContext turnContext, BotCallbackHandler callback, CancellationToken cancellationToken)
Microsoft.Teams.AI.TeamsAdapter: Information: SendActivitiesAsync for 1 activities.
Microsoft.Teams.AI.TeamsAdapter: Information: Sending activity.  ReplyToId: 1720538447636
Microsoft.Teams.AI.TeamsAdapter: Information: SendActivitiesAsync for 1 activities.
Microsoft.Teams.AI.TeamsAdapter: Information: Sending activity.  ReplyToId: 1720538447636
Microsoft.Teams.AI.TeamsAdapter: Information: SendActivitiesAsync for 1 activities.
Microsoft.Teams.AI.TeamsAdapter: Information: Sending activity.  ReplyToId: 1720538447636
Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker: Information: Executed action method RubrikTeamsApp.Controllers.BotController.PostAsync (RubrikTeamsApp), returned result Microsoft.AspNetCore.Mvc.EmptyResult in 3429.0032ms.
Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker: Information: Executed action RubrikTeamsApp.Controllers.BotController.PostAsync (RubrikTeamsApp) in 3447.1569ms
Microsoft.AspNetCore.Routing.EndpointMiddleware: Information: Executed endpoint 'RubrikTeamsApp.Controllers.BotController.PostAsync (RubrikTeamsApp)'
Microsoft.AspNetCore.Hosting.Diagnostics: Information: Request finished HTTP/1.1 POST http://localhost/api/messages - 200 0 - 3460.4155ms
The thread '.NET TP Worker' (20740) has exited with code 0 (0x0).
The thread '.NET TP Worker' (5932) has exited with code 0 (0x0).
The thread '.NET TP Worker' (42996) has exited with code 0 (0x0).
The thread '.NET TP Worker' (40636) has exited with code 0 (0x0).
The thread '.NET TP Worker' (48280) has exited with code 0 (0x0).
The thread '.NET TP Worker' (18476) has exited with code 0 (0x0).

```

I can see the response content,  but just will not output to Teams Bot.

my skprompt.txt I have this

```

The following is a conversation with an AI assistant. 
The assistant can turn a light on or off.
The assistant must return the following JSON structure:

{"type":"plan","commands":[{"type":"DO","action":"<name>","entities":{"<name>":<value>}},{"type":"SAY","response":"<response>"}]}

The following actions are supported:

- LightsOn
- LightsOff
- Pause time=<duration in ms>
- LightStatus

The lights are currently {{getLightStatus}}.

Always respond in the form of a JSON based plan. Stick with DO/SAY.

```

and config.json I have this


```

{
  "schema": 1.1,
  "description": "A bot that can turn the lights on and off",
  "type": "completion",
  "completion": {
    "model": "gpt-3.5-turbo",
    "completion_type": "chat",
    "include_history": true,
    "include_input": true,
    "max_input_tokens": 2800,
    "max_tokens": 1000,
    "temperature": 0.2,
    "top_p": 0.0,
    "presence_penalty": 0.6,
    "frequency_penalty": 0.0,
    "stop_sequences": []
  },
  "augmentation": {
    "augmentation_type": "sequence"
  }
}

```

**Comment by garrytrinder** [comment]
I'm also seeing similar behaviour to @packet3 using .NET SDK v1.4.0 and Azure OpenAI Service.

![image](https://github.com/user-attachments/assets/9eded481-c523-4320-82e6-42c77e62f340)

`SettingOn` and `SettingOff` actions work fine, but issue occur when using the Prompt Function.

If I use the exact phrase used in the example given in the `skprompt.txt` it works fine, but if I slightly deviate from it, I get the error.

Below are the files I am using.

## config.json

```json
{
  "schema": 1.1,
  "description": "AI Bot that can turn a setting on and off",
  "type": "completion",
  "completion": {
    "model": "gpt-3.5-turbo",
    "completion_type": "chat",
    "include_history": true,
    "include_input": true,
    "max_input_tokens": 2800,
    "max_tokens": 1000,
    "temperature": 0.2,
    "top_p": 0.0,
    "presence_penalty": 0.6,
    "frequency_penalty": 0.0,
    "stop_sequences": []
  },
  "augmentation": {
    "augmentation_type": "sequence"
  }
}
```

## actions.json

```json
[
  {
    "name": "SettingOn",
    "description": "Turns on the setting"
  },
  {
    "name": "SettingOff",
    "description": "Turns off the setting"
  }
]
```

## skprompt.json

```text
#### START PROMPT INSTRUCTIONS

## Prompt grounding
- You are an AI assistant that helps people find information. 
- Your name is Bob. 
- The Person is someone who wants to perform actions.

The following actions are supported:

- SettingOn
- SettingOff

The status of the setting is currently {{getSettingStatus}}.

#### END PROMPT INSTRUCTIONS

#### START ACTION INSTRUCTIONS 

What to do after action and functions have been executed: 
- Return the action results to the Person. 
- If the chat doesn't match the action then just use standard chat. 
- The Assistant must return the following JSON structure if it matches an action:
{"type":"plan","commands":[{"type":"DO","action":"<name>","entities":{"<name>":<value>}},{"type":"SAY","response":"<response>"}]}

#### END ACTION INSTRUCTIONS 

#### START ACTION EXAMPLES

Person: What is the status of the setting? 
Bob: The setting is currently {{getSettingStatus}}.

#### END ACTION EXAMPLES

{{$history}}
Person: {{$input}}
Bob: 
```

**Comment by donatas-xyz** [comment]
I've migrated my app from JavaScript to C# with `teams-ai 1.5.0` and gpt4, and I'm still facing the same issue: `Reached max model response repair attempts. Last feedback given to model: “Return a JSON object that uses the SAY command to say what you’re thinking.”`. 

Actions do work intermittently - just like in the JavaScript based app. So the code is fine - just the model cannot validate its own responses properly. 

FYI @aacebo @garrytrinder

**Comment by SLdragon** [comment]
I am getting the same issue in .NET, if I manually update the skprompt as below, it works:

```
The following is a conversation with an AI assistant.
The assistant can turn a light on or off.

context:
   {
      "name":"LightsOn",
      "description":"Turns on the lights"
   }
```

However, and if I use skprompt as below:

```
﻿The following is a conversation with an AI assistant.
The assistant can turn a light on or off.

context:

[
   {
      "name":"LightsOn",
      "description":"Turns on the lights"
   },
   {
      "name":"LightsOff",
      "description":"Turns off the lights"
   },
   {
      "name":"dummy",
      "description":"dummy"
   }
]
```
It will show error:

The bot encountered an unhandled error: Reached max model response repair attempts. Last feedback given to model: "The JSON returned had errors. Apply these fixes: "

**Comment by donatas-xyz** [comment]
I don't claim I've found a fix to this issue, but since this change, it only occurred to me once with JS teams-ai 1.3.1 and hasn't happened with C# teams-ai 1.5.0 yet. Previously it used to constantly fail. 

So I've noticed in someone's code on GitHub, that instead of using this JSON string in their instructions 
```json
{
	"type": "plan",
	"commands": [
		{
			"type": "DO",
			"action": "<name>",
			"entities": {
				"<name>": "<value>"
			}
		},
		{
			"type": "SAY",
			"response": "<response>"
		}
	]
}
```

they are using this one instead: 
```json
{
	"type": "plan",
	"commands": [
		{
			"type": "DO",
			"action": "<name>",
			"parameters": {
				"<name>": "<value>"
			}
		},
		{
			"type": "SAY",
			"response": "<response>"
		}
	]
}
```

Notice how instead of `entities` it's now `parameters`? I've changed the value and now LLM is as stable with its responses as anything. Tested with both GPT-3.5 and GPT-4.

Hopefully this will last and help someone else as well. 

FYI @dclaux @packet3 @SLdragon @garrytrinder @aacebo @simaonogueira101
